File: packages/engine-freecut/src/executor/planValidator.ts
// packages/engine-freecut/src/executor/planValidator.ts
import { Action } from "./types";
import { AssetResolver } from "./assetResolver";

export interface ValidationResult {
  ok: boolean;
  errors: string[];
  warnings: string[];
  mediaIds: string[];
}

export async function validatePlan(
  actions: Action[],
  resolver: AssetResolver
): Promise<ValidationResult> {
  const errors: string[] = [];
  const warnings: string[] = [];
  const mediaIds: string[] = [];

  if (!Array.isArray(actions)) {
    return { ok: false, errors: ["Plan is not an array"], warnings, mediaIds };
  }

  const declaredClips = new Map<string, string>(); // clipId -> trackId

  for (let i = 0; i < actions.length; i++) {
    const a = actions[i];
    const where = `actions[${i}] (${a?.type ?? "unknown"})`;

    switch (a.type) {
      case "addMedia": {
        if (!a.mediaId) errors.push(`${where}: missing mediaId`);
        if (!a.clipId) errors.push(`${where}: missing clipId`);
        if (typeof a.startTime !== "number")
          errors.push(`${where}: startTime must be a number`);
        if (declaredClips.has(a.clipId))
          warnings.push(`${where}: duplicate clipId "${a.clipId}"`);
        declaredClips.set(a.clipId, a.trackId);
        mediaIds.push(a.mediaId);
        break;
      }
      case "split": {
        if (!declaredClips.has(a.clipId))
          errors.push(`${where}: split references unknown clipId "${a.clipId}"`);
        if (typeof a.time !== "number" || a.time <= 0)
          errors.push(`${where}: split time must be > 0`);
        // After split, segment_2 will be addressable as `${clipId}_segment_2`
        declaredClips.set(`${a.clipId}_segment_1`, a.trackId);
        declaredClips.set(`${a.clipId}_segment_2`, a.trackId);
        break;
      }
      case "updateClip": {
        if (!declaredClips.has(a.clipId))
          errors.push(`${where}: updateClip references unknown clipId "${a.clipId}"`);
        const speed = a.properties?.playbackSpeed;
        if (speed !== undefined && (speed < 0.1 || speed > 4.0))
          errors.push(`${where}: playbackSpeed ${speed} out of range [0.1, 4.0]`);
        break;
      }
      case "addCaption": {
        if (!a.text) errors.push(`${where}: caption text required`);
        if (typeof a.duration !== "number" || a.duration <= 0)
          errors.push(`${where}: caption duration must be > 0`);
        break;
      }
      case "removeClip": {
        if (!declaredClips.has(a.clipId))
          warnings.push(`${where}: removeClip references unknown clipId "${a.clipId}"`);
        break;
      }
      default:
        errors.push(`${where}: unknown action type`);
    }
  }

  // verify all media exist
  const { unresolved } = await resolver.assertAllExist([...new Set(mediaIds)]);
  for (const u of unresolved) errors.push(`unresolved media: ${u}`);

  return { ok: errors.length === 0, errors, warnings, mediaIds };
}

File: monet/engines/freecut/executor/plan_validator.py
# monet/engines/freecut/executor/plan_validator.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import List, Dict

from .types import Action
from .asset_resolver import AssetResolver


@dataclass
class ValidationResult:
    ok: bool
    errors: List[str] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)
    mediaIds: List[str] = field(default_factory=list)


def validate_plan(actions: List[Action], resolver: AssetResolver) -> ValidationResult:
    errors: List[str] = []
    warnings: List[str] = []
    media_ids: List[str] = []
    declared_clips: Dict[str, str] = {}

    for i, a in enumerate(actions):
        where = f"actions[{i}] ({a.type})"

        if a.type == "addMedia":
            if not a.mediaId:
                errors.append(f"{where}: missing mediaId")
            if a.clipId in declared_clips:
                warnings.append(f'{where}: duplicate clipId "{a.clipId}"')
            declared_clips[a.clipId] = a.trackId
            media_ids.append(a.mediaId)

        elif a.type == "split":
            if a.clipId not in declared_clips:
                errors.append(f'{where}: unknown clipId "{a.clipId}"')
            if a.time <= 0:
                errors.append(f"{where}: split time must be > 0")
            declared_clips[f"{a.clipId}_segment_1"] = a.trackId
            declared_clips[f"{a.clipId}_segment_2"] = a.trackId

        elif a.type == "updateClip":
            if a.clipId not in declared_clips:
                errors.append(f'{where}: unknown clipId "{a.clipId}"')
            sp = a.properties.playbackSpeed
            if sp is not None and not (0.1 <= sp <= 4.0):
                errors.append(f"{where}: playbackSpeed {sp} out of range [0.1,4.0]")

        elif a.type == "addCaption":
            if not a.text:
                errors.append(f"{where}: text required")
            if a.duration <= 0:
                errors.append(f"{where}: duration must be > 0")

        elif a.type == "removeClip":
            if a.clipId not in declared_clips:
                warnings.append(f'{where}: removeClip unknown clipId "{a.clipId}"')

    resolved, unresolved = resolver.assert_all_exist(list(set(media_ids)))
    for u in unresolved:
        errors.append(f"unresolved media: {u}")

    return ValidationResult(ok=len(errors) == 0, errors=errors, warnings=warnings, mediaIds=media_ids)
